tsl/src/chunk_merge.c

```diff
@@ -828,22 +831,41 @@ chunk_merge_chunks(PG_FUNCTION_ARGS)
 		LOCKMODE lockmode =
 			(lock_upgrade == MERGE_LOCK_ACCESS_EXCLUSIVE) ? AccessExclusiveLock : ExclusiveLock;
 
-		rel = try_table_open(relid, lockmode);
+		chunk = ts_chunk_get_by_relid_locked(relid, lockmode, false);
 
 		/* Check if the table actually exists. If not, it could have been
-		 * deleted in a concurrent merge. */
-		if (rel == NULL)
-			ereport(ERROR,
-					(errcode(ERRCODE_UNDEFINED_TABLE),
-					 errmsg("relation does not exist"),
-					 errdetail("The relation with OID %u might have been removed "
-							   "by a concurrent merge or other operation.",
-							   relid)));
+		 * deleted in a concurrent merge or be another type of relation. */
+		if (chunk == NULL)
+		{
+			rel = try_relation_open(relid, AccessShareLock);
```

Comment:
Code will be significantly simpler if you use a switch (or normal if-statement, no opinion here) on result of `get_rel_relkind(relid)` here instead, since you don't need the relation open. It will return NUL character (`'\0'`) if the relation does not exist, and a proper relkind constant otherwise.

Reply:
I will clarify in a better comment.